### PR TITLE
Store link to open backorder

### LINKED ISSUE #12952: [DFC Orders] Store links to open backorders
#### :information_source: Funded Feature. Please track ALL ASSOCIATED WORK under the associated tracking code `#11678 DFC Orders`

Part of:

- #11678

When a customer places a new order, we search the remote orders endpoint for an open order to add to or create a new one. But when we find an open order, it could have been placed by another platform linked by the same user. We should store the link to an open backorder for several reasons:

* Avoid changing an order we didn't create.
* Know locally if we have a backorder already.

### PR DESCRIPTION
**Pull request by mkllnk**
#### What? Why?

- Closes #12952 

A new scenario is be possible now:

- Distributor has two order cycles open with products from the same DFC supplier.
- Previously, we would create only one backorder.
- The backorder got finalised at the end of the first order cycle.
- Now we create a backorder for each order cycle.
- The backorders are amended independently for each order cycle.
- The backorders get finalised at the end of their respective order cycle.

#### What should we test?

- Testing for @RaggedStaff:
- Open new order cycle with DFC products.
- Place orders which trigger a backorder.
- Cancel one order -> backorder should get adjusted as before.
- At OC end, the backorder should get finalised like before.

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [x] Technical changes only
- [ ] Feature toggled

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: app/services/fdc_backorderer.rb
Comment (rioug) [inline app/services/fdc_backorderer.rb:46]: typo:
```suggestion
  # DEPRECATED
```

File: app/services/fdc_backorderer.rb
Comment (rioug) [inline app/services/fdc_backorderer.rb:32]: TIL about `lazy`

**Review by rioug — APPROVED** [review]
Looks good ! Thanks for all the comments   on the `lookup_open_order` :1st_place_medal: I don't think I would have understood without them !

**Comment by RaggedStaff** [comment]
Heads up - this is not working!

**DO NOT MERGE**

File: app/jobs/backorder_job.rb
Comment (dacook) [inline app/jobs/backorder_job.rb:16]: 💎✨

File: app/services/fdc_backorderer.rb
Comment (dacook) [reply app/services/fdc_backorderer.rb:32]: Me too!

**Review by dacook — COMMENTED** [review]
Great work, a nice step-by-step approach which seems like it should work perfectly..

But apparently it didn't 😞

**Comment by mkllnk** [comment]
@RaggedStaff, do you have any more information on this?

You shared a couple of screenshots on Slack with me which state a completed order of:

- Baked Beans * 2
- Apple Vinegar * 2
- Camelina Seed * 1
- Carlin Peas * 1

I found two orders on the staging server that seem to be related. Both were cancelled. But they contain a mix of retail and wholesale items. I don't think we specified that scenario, did we?

Also, the resulting Shopify order doesn't have all the items of the two cancelled orders. So something must have gotten cancelled, right?

I also noticed that the `updated_at` timestamps of the orders are only 100ms apart. Did you cancel them both at the same time? I'm not sure we guard against race conditions during cancellations. 

Can you explain exactly what you did and what you expected? I'm wondering if the real pilots are actually simpler. We had some assumptions:

- Either stock controlled or on demand.
- Either retail or wholesale variants sold in OFN.
- Canellations are triggered by the customer and therefore rare and don't happen in bulk.

Some of this may be solved later when we support any way of amending orders but we are not there yet. This pull request just needs to work as well as before.

**Comment by mkllnk** [comment]
@RaggedStaff, I just did a little test on staging and everything seemed to work. Steps:

* Ordered beans and chia seed.
* Draft order open with bean while chia was still in stock.
* Second order with beans and chia.
* Draft order now contained chia as well but no additional beans, still in stock.
* Cancel second order.
* Draft order got reverted to first state of only having beans.
* At the end of the order cycle the draft order was completed with beans.

Can you check again? Maybe we can find the edge cases. I guess though that the edge cases never worked. This PR doesn't solve everything, only one little aspect that will allow more.

**Comment by RaggedStaff** [comment]
> @RaggedStaff, do you have any more information on this?
> 
> 
> I found two orders on the staging server that seem to be related. Both were cancelled. But they contain a mix of retail and wholesale items. I don't think we specified that scenario, did we?

No we didn't, the staging Shopify store is in a mess though, cos we keep switching back & forth to test different scenarios.
 
> Also, the resulting Shopify order doesn't have all the items of the two cancelled orders. So something must have gotten cancelled, right?

I think some stuff was in stock, so didn't ever appear on the order.
 
> I also noticed that the `updated_at` timestamps of the orders are only 100ms apart. Did you cancel them both at the same time? I'm not sure we guard against race conditions during cancellations.

Ah, yeah that may well be it. I'm pretty sure I used the bulk cancel Order function (selected both orders on the Order screen & cancelled). I think I expected that to cancel both orders and either:
1. Recalculate Producer backorder for new demand (summing up remaining orders)
2. Submit an adjustment for each cancellation in sequence

I don't think I'm bothered which way we go, it should result in the same outcome, right? 🤔 

I think the edge here may be an empty order... Shopify will not allow an Order without at least 1 OrderLine with hasQuantity > 0. Have you tried cancelling all the Orders & seeing what happens? I think that may explain my issue - the first order was left on the Shopify order (i.e. it failed attempting to back out the last orderline). 

I would like to confirm that as I'd like better error capture on that in Shopify (& possibly OFN). I'll do a bit more testing now.

If that is it, I agree the issue probably existed already & we don't need to address it in this PR. It's (hopefully) quite an unlikely edge case, at least for our current pilots. 🤞🏻

**Comment by RaggedStaff** [comment]
Ok. That's not the edge case... I checked with a single order & cancellation... steps to reproduce:

1. Set up up Test OC ([18](https://staging.openfoodnetwork.org.uk/admin/order_cycles/216857/edit)) on FDC Test Hub 2
2. Place Order for Red Fox Carlin Peas (Order [#R363653617](https://staging.openfoodnetwork.org.uk/admin/orders/R363653617/edit))
3. Confirm Draft Order is present on test-hodmedod (Draft [#D124](https://admin.shopify.com/store/test-hodmedod/draft_orders/1184779272499))
4. Cancel Order [#R363653617](https://staging.openfoodnetwork.org.uk/admin/orders/R363653617/edit)
5. Confirm Draft Shopify order ( [#D124](https://admin.shopify.com/store/test-hodmedod/draft_orders/1184779272499)) now only has placeholder product.

Seems it is multi-order cancellations. I will warn our test hubs not to attempt that & I think we can proceed with this now.

Thanks @mkllnk

**Comment by mkllnk** [comment]
> multi-order cancellations

Yes, we probably have multiple cancellations run at the same time going into a race condition. I would like to fix that a bit later though, when we amend everything consistently. Hm, maybe it doesn't matter when I fix this. Just a matter of priority.

**Comment by mkllnk** [comment]
@RaggedStaff, how shall we proceed here?

* The race condition of multiple cancellations at the same time is a pre-existing issue and can be fixed later.
* The error on completing the backorder was caused by an empty backorder which doesn't need placing anyway. But I added a commit to handle that case gracefully.

Do you approve of this PR? Can we merge it?

**Comment by mkllnk** [comment]
After chatting with David, we agree that we can merge this. It has been tested and even though the test failed, we know that this issues have been either addressed or were pre-existing. So let's unblock further work.

@RaggedStaff, if you have concerns then you can still opt out of the next release until we fix whatever your concern is.

**Pull request merged**
